### PR TITLE
add support for buf

### PR DESCRIPTION
**Pull request by remyleone**

### CODE REVIEW

**Comment by roidelapluie** [comment]
Thanks! I will review asap.

File: .github/workflows/buf.yml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  steps:
+    - uses: actions/checkout@v2
+    - uses: bufbuild/buf-setup-action@v0.5.0
+    - uses: bufbuild/buf-lint-action@v1
+      with:
+        input: 'prompb'
+    - uses: bufbuild/buf-breaking-action@v1
+      with:
+        input: 'prompb'
+        against: 'https://github.com/prometheus/prometheus.git#branch=main,ref=HEAD~1,subdir=prompb'
```
Comment (amckinney) [inline .github/workflows/buf.yml:16]: I don't think you need the `ref=HEAD~1` here - if you're comparing against the GitHub repository directly, the `main` branch in the `prompb` sub-directory should be sufficient, no?

Comment (amckinney) [reply .github/workflows/buf.yml:16]: Ah after revisiting [the documentation](https://docs.buf.build/ci-cd/github-actions#buf-push), I may have misspoke. This looks fine as-is.

Comment (remyleone) [reply .github/workflows/buf.yml:16]: I've used the example in https://github.com/bufbuild/buf-push-action/blob/main/README.md to stay as close as possible to the official documentation. Buf if the main branch is enough let's simplify it :)

**Comment by johanbrandhorst** [comment]
Note that the version of protos that `envoy` depends on is in https://github.com/prometheus/client_model. Adding this repository is nice but not a requirement for `envoy` 😬.

File: .github/workflows/buf.yml (hunk 1 above)
Comment (remyleone) [reply .github/workflows/buf.yml:16]: no problem I've revert it :)

File: .github/workflows/buf.yml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+on:
+  push:
+    branches:
+      - main
+
+jobs:
+  steps:
+    - uses: actions/checkout@v2
+    - uses: bufbuild/buf-setup-action@v0.5.0
```
Comment (amckinney) [inline .github/workflows/buf.yml:9]: We can use `v0.6.0` here:
```suggestion
    - uses: bufbuild/buf-setup-action@v0.6.0
```

I'll update the out-of-date README.md in `buf-push-action`.

**Comment by remyleone** [comment]
@johanbrandhorst should envoy be updated to use a different dependency than https://github.com/prometheus/client_model ? It seems it is deprecated. I'm just wondering if it is a good idea to push this schema as well.

**Comment by johanbrandhorst** [comment]
> @johanbrandhorst should envoy be updated to use a different dependency than https://github.com/prometheus/client_model ? It seems it is deprecated. I'm just wondering if it is a good idea to push this schema as well.

You'll have to take that up with the envoy maintainers, I expect it's there for historical reasons only. And sure, I think it'd still be great to have this published to the BSR of course ⭐.

**Comment by remyleone** [comment]
@adisuissa what do you think?

**Comment by remyleone** [comment]
https://github.com/prometheus/client_model/pull/50

**Comment by adisuissa** [comment]
> @adisuissa what do you think?

Looking at the code it seems that Envoy is using the `MetricFamily` type ([here](https://github.com/envoyproxy/envoy/blob/d3ef83a5cb822e0ba1a13e4c32abf6f061bdddf1/api/envoy/service/metrics/v3/metrics_service.proto#L53)), and the upgrade you are suggesting will require some changes in the Envoy code.

Another point, I don't think that the client_model is deprecated, and it seems Envoy is using this [commit](https://github.com/prometheus/client_model/commit/147c58e9608a4f9628b53b6cc863325ca746f63a), which is after the deprecation note.

**Comment by remyleone** [comment]
@adisuissa What kind of changes are required on envoy side to use this upgrade? Do you think those changes are substantial?

**Comment by johanbrandhorst** [comment]
> > @adisuissa what do you think?
> 
> Looking at the code it seems that Envoy is using the `MetricFamily` type ([here](https://github.com/envoyproxy/envoy/blob/d3ef83a5cb822e0ba1a13e4c32abf6f061bdddf1/api/envoy/service/metrics/v3/metrics_service.proto#L53)), and the upgrade you are suggesting will require some changes in the Envoy code.

I don't think it should be necessary to update envoy here, that sounds unnecessarily complicated. This repository is not a dependency of envoy.

**Comment by roidelapluie** [comment]
It seems someone else owns prometheus in buf.build, I am trying to figure it out.

**Comment by johanbrandhorst** [comment]
Hi @roidelapluie, as I said in the other PR, this org is currently managed by Buf but we're happy to hand it over to the Prometheus maintainers if they want to take ownership and responsibility of the modules. I'd be happy to discuss more on slack or a call, where is the best place to meet you?

**Comment by roidelapluie** [comment]
@johanbrandhorst do you want to review this?

File: .github/workflows/buf.yml [hunk 3]
```diff
@@ -0,0 +1,20 @@
+on:
+  push:
+    branches:
+      - main
```
Comment (johanbrandhorst) [inline .github/workflows/buf.yml:4]: I think you'd probably want two separate jobs - one that runs on `main` that does the push, and another that runs on every PR and does the other checks (lint, breaking). Maybe break this into two jobs?

File: prompb/buf.yaml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+version: v1
+name: buf.build/prometheus/prometheus
```
Comment (johanbrandhorst) [inline prompb/buf.yaml:2]: @roidelapluie this repository doesn't exist yet on the BSR, could you create it?

File: .github/workflows/buf.yml [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+      with:
+        input: 'prompb'
+    - uses: bufbuild/buf-breaking-action@v1
+      with:
+        input: 'prompb'
+        against: 'https://github.com/prometheus/prometheus.git#branch=main,ref=HEAD~1,subdir=prompb'
+    - uses: bufbuild/buf-push-action@v1
+      with:
+        input: 'prompb'
+        buf_token: ${{ secrets.BUF_TOKEN }}
```
Comment (johanbrandhorst) [inline .github/workflows/buf.yml:20]: @roidelapluie you'll need to add a token of yours to the secrets for this step to work.

**Review by johanbrandhorst — COMMENTED** [review]
I think it'd be best to split the workflow into two jobs - one that does CI level checks and one that does CD level pushes.

**Comment by roidelapluie** [comment]
Thanks @remyleone, let's see how this work and later implement dedicated linting.

**Pull request merged**